# Bài tập 5


## Câu 1 (1 điểm)

Dựa vào công thức: $$\mathbb{E}_{\mathcal{D}}[E_{in}(\mathbf{w}_{lin})] = \sigma^2\left(1 - \frac{d+1}{N}\right)$$

Tại $\sigma = 0.1$ và $d = 8$, ta tìm $N$ nhỏ nhất để $E_{in}$ lớn hơn $0.008$ bằng cách giải bất phương trình biến đổi từ công thức trên:

$$N > \frac{d+1}{1 - \frac{0.008}{\sigma^2}}$$

In [1]:
sigma = 0.1
d = 8
N = (d+1) / (1 - 0.008/sigma**2 )

print("N thỏa hệ phương trình khi: N > {}".format(round(N)))

N thỏa hệ phương trình khi: N > 45


YOUR ANSWER HERE Đáp án nhỏ nhất để thóa điều kiện trên là N=100. Do đó, em chọn đáp án [c] 100

## Câu 2 (1 điểm)

$$h(\mathbf{x})=sign(\tilde{w}_0 + \tilde{w}_1 x_1^2 + \tilde{w}_2 x_2^2)$$ 

- Khi $x_1$ siêu âm hoặc siêu dương thì $x_1^2$ sẽ dương và rất lớn, và  $\tilde{w}_1 x_1^2$ sẽ là đại lượng quyết định dấu của tổng trong công thức ở trên. Do $x_1^2$ dương nên $\tilde{w}_1$ sẽ quyết định dấu của $\tilde{w}_1 x_1^2$.
- Khi $x_2$ siêu âm hoặc siêu dương thì $x_2^2$ sẽ dương và rất lớn, và  $\tilde{w}_2 x_2^2$ sẽ là đại lượng quyết định dấu của tổng trong công thức ở trên. Do $x_2^2$ dương nên $\tilde{w}_2$ sẽ quyết định dấu của $\tilde{w}_2 x_2^2$.

**Dựa vào mô tả như hình, ta thấy:**

- Khi giá $x_1$ càng lớn (tiến về $-\infty$ hoặc $+\infty$) thì $h(\mathbf{x}) = -1$. Mà  $\tilde{w}_1$ quyết định dấu của $h(\mathbf{x})$ nên $\tilde{w}_1 < 0$
- Khi giá $x_2$ càng lớn (tiến về $-\infty$ hoặc $+\infty$) thì $h(\mathbf{x}) = +1$. Mà  $\tilde{w}_2$ quyết định dấu của $h(\mathbf{x})$ nên $\tilde{w}_2 > 0$

YOUR ANSWER HERE Do đó, em chọn đáp án [d]

## Câu 3 (1 điểm)

Dựa vào đề, ta thấy số đa thức có thể có khi triển khai $x_1$, $x_2$ trong không gian $R^2$ là 15.

Dựa vào công thức ở slide 4, lecture 9 là:

- $\tilde{d}_{vc} \le \tilde{d} + 1 = $ Số output của $\Phi_4 = 15$

=> 15 là giá trị đầu tiên **KHÔNG NHỎ HƠN** VC-dimension



YOUR ANSWER HERE Do đó, em chọn đáp án [c]

## Câu 4 (1 điểm)

Theo đề: 
$$E(u,v) = (ue^v - 2ve^{-u})^2$$

Đạo hàm riêng của $E(u,v)$ theo $u$:

$\displaystyle \frac{\partial E}{\partial u}$ $= 2*E*E' = 2*(ue^v - 2ve^{-u})* (e^u + 2ve^{-u})$

YOUR ANSWER HERE Do đó, em chọn đáp án [e]

## Câu 5 (1 điểm)

In [2]:
#Import thư viện cho tất cả các câu sau:
import math
import numpy as np

In [3]:
# Biến e
e = math.e


def E(u, v):
    return (u*e**v - 2*v*e**(-u))**2

#Code dùng cho câu 5,6:
'''
    input:  
            - eta
            - N: Số vòng lặp tối đa 
            - error: Độ lỗi yêu cầu
    output: 
            - result_i: Sô vòng lặp cần thiết để E(u,v) < error
            - result_x: Giá trị của [u,v] tại lần lặp thứ result_i
'''
def get_iter(eta,N,error):
    # Khai báo iter ban đầu = 0:
    result_i = 0
    # Biến e
    e = math.e
    # Bắt đầuv với  (u,v) = (1,1)
    x = [1,1]

    for i in range(1,N):
        u,v = x
        #Đạo hàm theo u:
        dE_du = 2*(u*e**v - 2*v*e**(-u))*(e**v + 2*v*e**(-u))
        #Đạo hàm theo v:
        dE_dv = 2*(u*e**v - 2*v*e**(-u))*(u*e**v - 2*e**(-u))
        grad = np.array([dE_du, dE_dv])

        #Cập nhật x:
        x = x - eta * grad
        
        #Lưu tạm i và x:
        result_i = i
        result_x = x
        
        #Điều kiện dừng khi E(u,v) < error đầu vào:
        if E(x[0], x[1]) < error:
            break
            
    return result_i, result_x

In [4]:
result_i, result_uv = get_iter(eta = 0.1,N = 10000,error = 10**(-14))
print("Tốn {} lần lặp để độ lỗi của E(u,v) < 10^-14".format(result_i))

Tốn 10 lần lặp để độ lỗi của E(u,v) < 10^-14


YOUR ANSWER HERE Do đó, em chọn đáp án [d]

## Câu 6 (1 điểm)

In [5]:
print("Giá trị của (u,v) tại câu 5 là {}".format(np.round(result_uv, 3)))

Giá trị của (u,v) tại câu 5 là [0.045 0.024]


YOUR ANSWER HERE Do đó, em chọn đáp án [e]

## Câu 7 (1 điểm)

In [6]:
def get_error(eta,N_iters):
    # Khai báo iter ban đầu = 0:
    result_i = 0
    # Biến e
    e = math.e
    # Bắt đầuv với  (u,v) = (1,1)
    x = [1,1]

    for i in range(N_iters):
        #Đạo hàm theo u:
        u,v = x
        dE_du = 2*(u*e**v - 2*v*e**(-u))*(e**v + 2*v*e**(-u))
        #Dọc theo tọa độ u để giảm lỗi
        grad = np.array([dE_du, 0])
        x = x - eta * grad
        
        #Đạo hàm theo v:
        u,v = x
        dE_dv = 2*(u*e**v - 2*v*e**(-u))*(u*e**v - 2*e**(-u))
        #Dọc theo tọa độ v để giảm lỗi
        grad = np.array([0, dE_dv])
        x = x - eta * grad

        result_i = i
        result_x = x
            
    return E(x[0], x[1])

In [7]:
eta = 0.1
N_iters = 15
print("Độ lỗi E(u,v) sau {} lần lặp bằng {}".format(N_iters, get_error(eta,N_iters)))

Độ lỗi E(u,v) sau 15 lần lặp bằng 0.13981379199615324


YOUR ANSWER HERE Kết quả $0.1398$ gần với $10^{-1}$ nhất. Do đó, em chọn đáp án [a]

## Câu 8 (1 điểm)

In [8]:
def generate_target_w():
    """
    Generates target_w (the vector of parameters of f) 
    from two random, uniformly distributed points in [-1, 1] x [-1, 1].
    
    Returns
    -------
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    """
    # Generate two points from a uniform distribution over [-1, 1]x[-1, 1]
    p1 = np.random.uniform(-1, 1, 2)
    p2 = np.random.uniform(-1, 1, 2)
    # Compute the target W from these two points
    target_w = np.array([p1[1]*p2[0] - p1[0]*p2[1], p2[1] - p1[1], p1[0] - p2[0]]).reshape((-1, 1))
    
    return target_w

In [9]:
def generate_data(N, d, target_w):

    """
    Generates a data set by generating random inputs and then using target_w to generate the 
    corresponding outputs.
    
    Parameters
    ----------
    N : int
        The number of examples.
    target_w : numpy array, shape (3, 1) 
        The vector of parameters of f.
    
    Returns
    -------
    X : numpy array, shape (N, 3)
        The matrix of input vectors (each row corresponds to an input vector); the first column of 
        this matrix is all ones.
    Y : numpy array, shape (N, 1)
        The vector of outputs.        
    """
    bad_data = True # `bad_data = True` means: data contain points on the target line 
                    # (this rarely happens, but just to be careful)
                    # -> y's of these points = 0 (with np.sign); 
                    #    we don't want this (y's of data must be -1 or 1)
                    # -> re-generate data until `bad_data = False`
    
    while bad_data == True:
        X = np.random.uniform(-1, 1, (N, d))
        X = np.hstack((np.ones((N, 1)), X)) # Add 'ones' column
        Y = np.sign(np.dot(X, target_w))
        if (0 not in Y): # Good data
            bad_data = False
    
    return X, Y


In [10]:
def get_w_n_epochs(X,Y):
    learning_rate = 0.01
    n_epochs = 0
    N = len(X)
    d = len(X[0])-1
    w = np.zeros(d+1)
    previous_w = np.copy(w)
    
    while True:
            previous_w = np.copy(w)
            rand_idxs = np.random.permutation(N)
            for idx in rand_idxs:
                x = X[idx,:]
                y = Y[idx]
                gradient = -y * x / (1 + math.exp(y * np.dot(w.T, x)))
                w = w - learning_rate * gradient

            n_epochs += 1 
            if np.linalg.norm(w - previous_w) < 0.01:
                break
    return w, n_epochs

In [11]:
#Tính w cho câu dưới:
N = 100
d=2
generate_w = generate_target_w()
X,Y = generate_data(N, d, generate_w)

w, n_epochs = get_w_n_epochs(X,Y)

In [12]:
N_test = 100
X,Y = generate_data(N_test, d, generate_w)
E_out_total = 0       

# Calculate E_out via cross entropy error
E_out = []
epochs = []
for i in range(N_test):
    temp = np.log(1 + np.exp(-Y[i] * w.dot(X[i,:])))
    w_ ,n_epochs = get_w_n_epochs(X,Y)
    E_out.append(temp)
    epochs.append(n_epochs)
    

E_out_total = np.mean(E_out)

print("E_out = {}".format(E_out_total))

E_out = 0.13206908043477397


YOUR ANSWER HERE Giá trị của E_out gần với 0.1 nhất. Do đó, em chọn đáp án [d]

## Câu 9 (1 điểm)

In [13]:
print("epochs = {}".format(np.mean(epochs)))

epochs = 379.09


YOUR ANSWER HERE Kết quả gần với giá trị 350 nhất. Do đó, em chọn đáp án [a]

## Câu 10 (1 điểm)

- PLA: 
    - Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ được phân lớp đúng thì $\mathbf{w}$ không thay đổi.
    - Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ bị phân lớp sai thì cập nhật: $\mathbf{w} \leftarrow \mathbf{w} + y^{(n)}\mathbf{x}^{(n)}$.

- SGD: với mỗi mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ thì đều cập nhật $\mathbf{w} \leftarrow \mathbf{w} - \alpha \times \mathbf{\nabla_wE_{in}}$.

Chọn $E_{in} = e(w) = -y\mathbf{w^Tx}$ cho SGD trên.

=> $\nabla e = -y\mathbf{x}$

Hàm cập nhật $w$ của SGD lúc này:  $\mathbf{w} \leftarrow \mathbf{w} - \alpha \times(-y\mathbf{x}) $

hay $\mathbf{w} \leftarrow \mathbf{w} + \alpha \times(y\mathbf{x})$

Tại $\alpha = 1$:  $\mathbf{w} \leftarrow \mathbf{w} + y\mathbf{x}$

Ta chỉ chọn các điểm bị phân loại sai trong PLA nên: $e(\mathbf{w}) = -min(0, -y\mathbf{w^Tx})$

(Các điểm được phân loại chính xác thì độ dốc (gradient) bằng 0 nên khi cập nhật thì không làm thay đổi $\mathbf{w}$)

YOUR ANSWER HERE Do đó, em chọn đáp án [e]